File: CSE305/achrysan_HW3/Java/hw3.java

import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Stack;
import java.io.*;

public class hw3 {
	
	private static Stack myStack = new Stack();
	private static HashMap myMap = new HashMap();
	public static void hw3(String input, String output){
		try {
			
		BufferedWriter write = new BufferedWriter(new FileWriter(output));
		BufferedReader read = new BufferedReader(new FileReader(input));
		String line;
		while((line = read.readLine()) != null){
			line = line.replace("\n","");
			System.out.println("The line is: "+line);
			if(line.length()>3 && !line.equals(":true:") && !line.equals(":false:") && !line.equals(":error:")  && !line.equals("swap") && !line.equals("equal") && !line.equals("bind") && !line.equals("lessThan") && !line.equals("quit")){
			if(line.substring(0, 4).equals("push")){
				try{
					String value = line.substring(5,  line.length());
					if(value.charAt(0) == '\"' && value.charAt(value.length()-1) == '\"'){
						value = value.substring(1, value.length()-1);
						value = value +'`';
						myStack.push(value);
					}
					else if(Character.isLetter(value.charAt(0))){
						myStack.push(value);
					}else{
				myStack.push(Integer.parseInt(line.substring(5, line.length())));
					}
				System.out.println(line.substring(5, line.length())+" is being pushed onto the stack.");
				System.out.println("The Stack now contains:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
				}catch(Exception e){
					myStack.push(":error:");
					System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
				}
			}else{ myStack.push(":error:");
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");}
			}else if(line.equals("pop")){
				try{
				myStack.pop();
				System.out.println("An element is being popped off!");
				}catch(Exception e){
					myStack.push(":error:");
				}
			}else if(line.equals(":true:") ||line.equals(":false:")||line.equals(":error:")){
				myStack.push(line);
				System.out.println(line +" is being pushed onto the stack.");
				System.out.println("The Stack now contains:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else if(line.equals("add")){
				add_vals();
			}else if(line.equals("sub")){
				sub_vals();
			}else if(line.equals("mul")){
				mul_vals();
			}else if(line.equals("div")){
				div_vals();
			}else if(line.equals("rem")){
				rem_vals();
			}else if(line.equals("neg")){
				neg_vals();
			}else if(line.equals("swap")){
				swap_vals();
			}else if(line.equals("and")){
				comp_and();
			}else if(line.equals("or")){
				comp_or();
			}else if(line.equals("not")){
				comp_not();
			}else if(line.equals("equal")){
				comp_equal();
			}else if(line.equals("lessThan")){
				comp_lessThan();
			}else if(line.equals("bind")){
				comp_bind();
			}else if(line.equals("if")){
				comp_if();
			}else if(line.equals("let")){
				myStack.push("xletx");
			}else if(line.equals("end")){
				Object topval = myStack.peek();
				myStack.pop();
				while(myStack.peek() != "xletx"){
					myStack.pop();
				}
				myStack.pop();
				myStack.push(topval);
			}else if(line.equals("quit")){
				System.out.println("This is what is being printed to the File: ");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
				for(int i = 0; i<myStack.size(); i++){
					if(myStack.get(i) instanceof String){
						String item = (String) myStack.get(i);
						if(item.charAt(item.length()-1) == '`'){
							item = item.substring(0, item.length()-1);
							myStack.remove(i);
							myStack.insertElementAt(item, i);
						}else{
							continue;
						}
					}
				}
				Iterator<Object> iter = myStack.iterator();
				for(int i = myStack.size()-1; i>=0; i--){
					write.write(myStack.get(i).toString());
					write.newLine();
				}
				
				//System.exit(0);
			}else{
				myStack.push(":error:");
				}
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
		read.close();
		write.close();
	} catch (IOException e) {
		System.out.println("Error Occured:\n");
		e.printStackTrace();
	}
		
	}

	private static void comp_if() {
		if(myStack.size() > 2){
			Object temp0 = myStack.peek();
			Object temp1 = myStack.get(myStack.size()-2);
			Object myVal_1 = myStack.peek();
			Object myVal_2 = myStack.get(myStack.size()-2);
			Object myVal_3 = myStack.get(myStack.size()-3);
			System.out.println("myval2: "+myVal_2);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
			if(myMap.containsKey(myVal_3)){
				myVal_3 = myMap.get(myVal_3);
			}
			if(myVal_3.equals(":true:") || myVal_3.equals(":false:")){
				if(myVal_3.equals(":true:")){
					myStack.pop();
					myStack.pop();
					myStack.pop();
					myStack.push(temp0);
				}else if(myVal_3.equals(":false:")){
					myStack.pop();
					myStack.pop();
					myStack.pop();
					myStack.push(temp1);
				}
			}else{
				myStack.push(":error:");
			}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void comp_bind() {
		if(myStack.size() > 1){
			if(myStack.get(myStack.size()-2) instanceof String){
				String x = (String) myStack.get(myStack.size()-2);
				if(x.charAt(x.length()-1) == '`' || x.equals(":error:") || x.equals("xletx") || x.equals(":true:") || x.equals(":false:") || x.equals(":unit:") || myStack.peek().equals(":error:")){
					myStack.push(":error:");
				}else{
					myMap.put(x, myStack.peek());
					Object origVal = myMap.get(x);
					System.out.println("x: "+x);
					if(myMap.containsKey(x)){
						if(myMap.containsKey(origVal)){
							myMap.put(x, myMap.get(origVal));
						}
					}
					myStack.pop();
					myStack.pop();
					myStack.push(":unit:");
				}
			}else{
				myStack.push(":error:");
			}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void comp_lessThan() {
		if(myStack.size() >1){
		Object myVal_1 = myStack.peek();
		Object myVal_2 = myStack.get(myStack.size()-2);
		System.out.println("myVak_1: "+myVal_1);
		System.out.println("myVak_2: "+myVal_2);
		if(myMap.containsKey(myVal_1)){
			myVal_1 = myMap.get(myVal_1);
		}
		if(myMap.containsKey(myVal_2)){
			myVal_2 = myMap.get(myVal_2);
		}
		if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
			if((int)myVal_2 < (int)myVal_1){
				myStack.pop();
				myStack.pop();
				myStack.push(":true:");
			}else{
				myStack.pop();
				myStack.pop();
				myStack.push(":false:");
			}
		}else{
			myStack.push(":error:");
		}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void comp_equal() {
		if(myStack.size() > 1){
			Object myVal_1 = myStack.peek();
			Object myVal_2 = myStack.get(myStack.size()-2);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
			if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
				if((int)myVal_1 == (int)myVal_2){
					
					myStack.pop();
					myStack.pop();
					myStack.push(":true:");
				}else{
					myStack.pop();
					myStack.pop();
					myStack.push(":false:");
				}
			}else{
				myStack.push(":error:");
			}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void comp_not() {
		if(myStack.size() > 0){
		Object myVal_1 = myStack.peek();
		if(myMap.containsKey(myVal_1)){
			myVal_1 = myMap.get(myVal_1);
		}
		if(!(myVal_1.equals(":true:")) && !(myVal_1.equals(":false:"))){
			myStack.push(":error:");
		}else{
			if(myVal_1.equals(":true:")){
				myStack.pop();
				myStack.push(":false:");
			}else if(myVal_1.equals(":false:")){
				myStack.pop();
				myStack.push(":true:");
			}
		}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void comp_or() {
		if(myStack.size() > 1){
			Object myVal_1 = myStack.peek();
			Object myVal_2 = myStack.get(myStack.size()-2);
				if(myMap.containsKey(myVal_1)){
					myVal_1 = myMap.get(myVal_1);
				}
				if(myMap.containsKey(myVal_2)){
					myVal_2 = myMap.get(myVal_2);
				}
			
				if(!(myVal_1.equals(":true:")) && !(myVal_1.equals(":false:")) || !(myVal_2.equals(":true:")) && !(myVal_2.equals(":false:"))){
					myStack.push(":error:");
				}else{
					if(myVal_1.equals(myVal_2)){
						Object temp = myVal_1;
						myStack.pop();
						myStack.pop();
						myStack.push(temp);
					}else if(myVal_1.equals(":true:") || myVal_2.equals(":true:")){
						myStack.pop();
						myStack.pop();
						myStack.push(":true:");
					}
				}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void comp_and() {
		if(myStack.size() > 1){
			Object myVal_1 = myStack.peek();
			Object myVal_2 = myStack.get(myStack.size()-2);
				if(myMap.containsKey(myVal_1)){
					myVal_1 = myMap.get(myVal_1);
				}
				if(myMap.containsKey(myVal_2)){
					myVal_2 = myMap.get(myVal_2);
				}
			
				if(!(myVal_1.equals(":true:")) && !(myVal_1.equals(":false:")) || !(myVal_2.equals(":true:")) && !(myVal_2.equals(":false:"))){
					myStack.push(":error:");
				}else{
					if(myVal_1.equals(myVal_2)){
						Object temp = myVal_1;
						myStack.pop();
						myStack.pop();
						myStack.push(temp);
					}else if(myVal_1.equals(":false:") || myVal_2.equals(":false:")){
						myStack.pop();
						myStack.pop();
						myStack.push(":false:");
					}
				}
		}else{
			myStack.push(":error:");
		}
		
	}

	private static void swap_vals() {
		System.out.println("swapping");
		if(myStack.size()>1){
			Object temp = myStack.pop();
			Object temp2 = myStack.pop();
			myStack.push(temp);
			myStack.push(temp2);
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
	}

	private static void neg_vals() {
		System.out.println("negating");
		if(myStack.size()>0){
			Object myVal_1 = myStack.peek();
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myVal_1 instanceof Integer){
				Object myVal1 = myStack.pop();
				if(myMap.containsKey(myVal1)){
					myVal1 = myMap.get(myVal1);
					//myVal_1 = Integer.parseInt(myVal_1);
				}
			int	result = 0 - ((int)myVal1);
			System.out.println("result: "+result);
			myStack.push(result);
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
		
	}

	private static void rem_vals() {
		System.out.println("modulating");
		if(myStack.size()>1){
			Object myVal_1 = myStack.peek();
			Object temp = myStack.pop();
			Object myVal_2 = myStack.peek();
			myStack.push(temp);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
			if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
				Object myVal1 = myStack.pop();
				Object myVal2 = myStack.pop();
				if(myMap.containsKey(myVal1)){
					myVal1 = myMap.get(myVal1);
					//myVal_1 = Integer.parseInt(myVal_1);
				}
				if(myMap.containsKey(myVal2)){
					myVal2 = myMap.get(myVal2);
				}
			int	result = ((int) myVal2) % ((int)myVal1);
			System.out.println("result: "+result);
			myStack.push(result);
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
		
	}

	private static void div_vals() {
		System.out.println("dividing");
		if(myStack.size()>1){
			Object myVal_1 = myStack.peek();
			Object temp = myStack.pop();
			Object myVal_2 = myStack.peek();
			myStack.push(temp);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
			if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
				Object myVal1 = myStack.pop();
				Object myVal2 = myStack.pop();
				if(myMap.containsKey(myVal1)){
					myVal1 = myMap.get(myVal1);
					//myVal_1 = Integer.parseInt(myVal_1);
				}
				if(myMap.containsKey(myVal2)){
					myVal2 = myMap.get(myVal2);
				}
			int	result = ((int) myVal2) / ((int)myVal1);
			System.out.println("result: "+result);
			myStack.push(result);
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
		
	}

	private static void mul_vals() {
		System.out.println("multiplying");
		if(myStack.size()>1){
			Object myVal_1 = myStack.peek();
			Object temp = myStack.pop();
			Object myVal_2 = myStack.peek();
			myStack.push(temp);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
			if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
				Object myVal1 = myStack.pop();
				Object myVal2 = myStack.pop();
				if(myMap.containsKey(myVal1)){
					myVal1 = myMap.get(myVal1);
					//myVal_1 = Integer.parseInt(myVal_1);
				}
				if(myMap.containsKey(myVal2)){
					myVal2 = myMap.get(myVal2);
				}
			int	result = ((int) myVal2) * ((int)myVal1);
			System.out.println("result: "+result);
			myStack.push(result);
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
		
	}

	private static void sub_vals() {
		System.out.println("subtracting");
		if(myStack.size()>1){
			Object myVal_1 = myStack.peek();
			Object temp = myStack.pop();
			Object myVal_2 = myStack.peek();
			myStack.push(temp);
			System.out.println("myVal_1: "+myVal_1);
			System.out.println("myVal_2: "+myVal_2);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
			if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
				Object myVal1 = myStack.pop();
				Object myVal2 = myStack.pop();
				if(myMap.containsKey(myVal1)){
					myVal1 = myMap.get(myVal1);
					//myVal_1 = Integer.parseInt(myVal_1);
				}
				if(myMap.containsKey(myVal2)){
					myVal2 = myMap.get(myVal2);
				}
			int	result = ((int) myVal2) - ((int)myVal1);
			System.out.println("result: "+result);
			myStack.push(result);
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
			}else{
				myStack.push(":error:");
				System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
			}
		
		
	}

	private static void add_vals() {
		System.out.println("adding");
		if(myStack.size()>1){
			Object myVal_1 = myStack.peek();
			Object temp = myStack.pop();
			Object myVal_2 = myStack.peek();
			myStack.push(temp);
			if(myMap.containsKey(myVal_1)){
				myVal_1 = myMap.get(myVal_1);
			}
			if(myMap.containsKey(myVal_2)){
				myVal_2 = myMap.get(myVal_2);
			}
		if(myVal_1 instanceof Integer && myVal_2 instanceof Integer){
			Object myVal1 = myStack.pop();
			Object myVal2 = myStack.pop();
			if(myMap.containsKey(myVal1)){
				myVal1 = myMap.get(myVal1);
				//myVal_1 = Integer.parseInt(myVal_1);
			}
			if(myMap.containsKey(myVal2)){
				myVal2 = myMap.get(myVal2);
			}
		int	result = ((int) myVal2) + ((int)myVal1);
		System.out.println("result: "+result);
		myStack.push(result);
		System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
		}else{
			myStack.push(":error:");
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
		}
		}else{
			myStack.push(":error:");
			System.out.print(myStack.toString().replaceAll("\\[", "").replaceAll("]", "")+"\n"+"\n");
		}
		
	}

}

File: CSE305/achrysan_HW3/Python/hw3.py
#!/usr/bin/env python
import sys
import copy
myStack = list();
myMap = dict();
temp = list();
envHolder = list();
global counter
counter = 0
def hw3(inString, outString):
    file_read = open(inString, 'r')
    text_in_f = file_read.readlines()
    print(len(text_in_f))
    file_write = open(outString, 'w')
    for line in text_in_f:
        line = line[:-1]
        print("The line: \""+line+"\"")
        if line[:4] == "push":
            value = line[5:len(line)]
            try:
                if value[0] == "\"" and value[len(value)-1] == "\"":
                    value = value[1:len(value)-1]
                    value = value + '`'
                    push_val(value)
                elif value[0].isalpha():
                    push_val(value)
                elif '.' in line:
                    push_val(":error:")
                else:
                    push_val(int(value))
            except:
                push_val(":error:")

        elif line == "pop":
            pop_val()
        elif line == ":true:" or line == ":false:" or line == ":error:":
            push_val(line)
        elif line == "add":
            add_vals()
        elif line == "sub":
            sub_vals()
        elif line == "mul":
            mul_vals()
        elif line == "div":
            div_vals()
        elif line == "rem":
            rem_vals()
        elif line == "neg":
            neg_val()
        elif line == "swap":
            swap_vals()
        elif line == "and":
            comp_and()
        elif line == "or":
            comp_or()
        elif line == "not":
            comp_not()
        elif line == "equal":
            comp_equal()
        elif line == "lessThan":
            comp_lessThan()
        elif line == "bind":
            bind()
        elif line == "if":
            comp_if()
        if line == "let":
            push_val("xletx")
        elif line == "end":
            topval = top_val(myStack)
            pop_val()
            while myStack[0]!= "xletx":
                pop_val()
            pop_val()
            push_val(topval)
        elif line == "qui" or line == "quit":
            print("\nThis is what's being written to the file:")
            print(*myStack, sep="\n")
            for item in myStack:
                if isinstance(item, str):
                    if item[len(item)-1] == '`':
                        item = item[0:len(item)-1]
                    else:
                        pass
                file_write.write("%s\n" % item)
            sys.exit()

def push_val(myVal):
    """if((not(isinstance(myVal,int))) and myVal!= ":true:" or myVal!= ":false:" or myVal != ":error:"):
        myStack.insert(0, ":error:")
        return myStack
    if(isinstance(myVal,int):
       myStack.insert("""
    myStack.insert(0,myVal)
    print(myVal,"is being pushed onto the stack")
    print(*myStack, sep="\n")
    print("")
    return myStack

def pop_val():
    if len(myStack) != 0:
        print(myStack[0],"is being pop'd off the stack")
        myStack.remove(myStack[0])
        return myStack
    else:
        push_val(":error:")
def top_val(myStack):
    if len(myStack) != 0:
        return myStack[0]
    else:
        pass
    
def add_vals():
    print("adding")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int):
            result = myVal_2 + myVal_1
            print("result: ",result)
            pop_val()
            pop_val()
            push_val(result)
        else:
            push_val(":error:")
    else:
        push_val(":error:")


def sub_vals():
    print("subtracting")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int):
            result = myVal_2 - myVal_1
            print("result: ",result)
            pop_val()
            pop_val()
            push_val(result)
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def mul_vals():
    print("multiplying")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int):
            result = myVal_2 * myVal_1
            print("result: ",result)
            pop_val()
            pop_val()
            push_val(result)
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def div_vals():
    print("dividing")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int) and myVal_2 != 0:
            result = myVal_2 // myVal_1
            print("result: ",result)
            pop_val()
            pop_val()
            push_val(result)
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def rem_vals():
    print("modulating")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int) and myVal_2 != 0:
            result = myVal_2 % myVal_1
            print("result: ",result)
            pop_val()
            pop_val()
            push_val(result)
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def neg_val():
    print("negating")
    if len(myStack)>0:
        myVal_1 = myStack[0]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        if isinstance(myVal_1,int):
            result = 0 - myVal_1
            pop_val()
            push_val(result)
            print("result: ",result)
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def swap_vals():
    print("swapping")
    if len(myStack)>1:
        temp = myStack[0]
        myStack[0] = myStack[1]
        myStack[1] = temp
        print(*myStack, sep="\n")
    else:
        push_val(":error:")
        print(*myStack, sep="\n")

def comp_and():
    print("checking -and-")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if myVal_1 != ":true:" and myVal_1 != ":false:" or myVal_2 != ":true:" and myVal_2 != ":false:":
            push_val(":error:")
        else:
            if myVal_1 == myVal_2:
                temp = myVal_1
                pop_val()
                pop_val()
                push_val(temp)
            elif myVal_1 == ":false:" or myVal_2 == ":false:":
                pop_val()
                pop_val()
                push_val(":false:")
    else:
        push_val(":error:")

def comp_or():
    print("checking -or-")
    if len(myStack)>1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if myVal_1 != ":true:" and myVal_1 != ":false:" or myVal_2 != ":true:" and myVal_2 != ":false:":
            push_val(":error:")
        else:
            if myVal_1 == myVal_2:
                temp = myVal_1
                pop_val()
                pop_val()
                push_val(temp)
            elif myVal_1 == ":true:" or myVal_2 == ":true:":
                pop_val()
                pop_val()
                push_val(":true:")
    else:
        push_val(":error:")

def comp_not():
    print("computing not")
    if len(myStack)>0:
        myVal_1 = myStack[0]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        if myVal_1 != ":true:" and myVal_1 != ":false:":
            push_val(":error:")
        else:
            if myVal_1 == ":true:":
                pop_val()
                push_val(":false:")
            elif myVal_1 == ":false:":
                pop_val()
                push_val(":true:")
    else:
        push_val(":error:")

def comp_equal():
    print("computing equal")
    if len(myStack) > 1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int):
            if myVal_1 == myVal_2:
                pop_val()
                pop_val()
                push_val(":true:")
            else:
                pop_val()
                pop_val()
                push_val(":false:")
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def comp_lessThan():
    print("computing lessThan")
    if len(myStack) > 1:
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        if isinstance(myVal_1,int) and isinstance(myVal_2,int):
            if myVal_2 < myVal_1:
                pop_val()
                pop_val()
                push_val(":true:")
            else:
                pop_val()
                pop_val()
                push_val(":false:")
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def bind():
    print("binding")
    if len(myStack) > 1:
        if isinstance(myStack[1], str):
            x = myStack[1]
            if x[len(x)-1] == "`" or x == ":error:" or x == "xletx" or x == ":true:" or x == ":false:" or x == ":unit:" or myStack[0] == ":error:":
                push_val(":error:")
            
            else:
                myMap[myStack[1]] = myStack[0]
                if x in myMap:
                    print("key: "+x)
                    if myMap[x] in myMap:
                        print("value: "+myMap[x])
                        print("value of value: "+str(myMap[myMap[x]]))
                        myMap[x] = myMap[myMap[x]]
                print (str(myStack[1])+" :" , str(myMap[myStack[1]]))
                pop_val()
                pop_val()
                push_val(":unit:")
        else:
            push_val(":error:")
    else:
        push_val(":error:")

def comp_if():
    if len(myStack) > 2:
        temp0 = myStack[0]
        temp1 = myStack[1]
        myVal_1 = myStack[0]
        myVal_2 = myStack[1]
        myVal_3 = myStack[2]
        for x in myMap:
            if x == myVal_1:
                myVal_1 = myMap[myVal_1]
        for x in myMap:
            if x == myVal_2:
                myVal_2 = myMap[myVal_2]
        for x in myMap:
            if x == myVal_3:
                myVal_3 = myMap[myVal_3]
        if myVal_3 == ":true:" or myVal_3 == ":false:":
            if myVal_3 == ":true:":
                pop_val()
                pop_val()
                pop_val()
                push_val(temp0)
            elif myVal_3 == ":false:":
                pop_val()
                pop_val()
                pop_val()
                push_val(temp1)
        else:
            push_val(":error:")
    else:
        push_val(":error:")
